core/src/main/java/bisq/core/api/CoreWalletsService.java

```diff
@@ -185,6 +196,37 @@ void sendBsq(String address,
         }
     }
 
+    TxFeeRateInfo getTxFeeRate() {
+        try {
+            CompletableFuture<Void> feeRequestFuture = CompletableFuture.runAsync(feeService::requestFees);
+            feeRequestFuture.get();  // Block until async fee request is complete.
```

Comment:
Is it really a good idea to block here? As discussed before, could we use a callback to avoid blocking?

Reply:
I'm not sure what the difference would be between future.get() and using a callback.   The wait for a response has to happen somewhere before `GrpcWalletsService` returns the gRPC reply.

Do you think `FeeService#requestFees` should be refactored to take a callback parameter?  Were you and chimp wanting to  refactor FeeService#requestFees?  Or should that be left alone, and make adjustments only in the api?

Another way would be to deal with the result handling is do it in the `GrpcWalletsService`.  I try to keep those boilerplate `Grpc*Service` classes as small as possible, but sometimes dealing with async calls in that layer is unavoidable.

If we are concerned about blocking forever, we could block with a timeout:  
`feeRequestFuture.get(5, SECONDS);  // Block until async fee request is complete.`

Reply:
I was thinking something like, in CoreWalletService
```
    TxFeeRateInfo getTxFeeRate(ResultHandler resultHandler, ErrorMessageHandler errorMessageHandler) {
            CompletableFuture<Void> feeRequestFuture = CompletableFuture.runAsync(feeService::requestFees);
            Futures.addCallback(feeRequestFuture, new FutureCallback<>() {

                @Override
                public void onSuccess(@Nullable Object result) {
                    resultHandler.handleResult();
                }

                @Override
                public void onFailure(Throwable t) {
                    errorMessageHandler.handleErrorMessage(t.getMessage());
                }
            }, MoreExecutors.directExecutor());
    }
```
where the resultHandler is doing the gRPC reply, something like
```
    @Override
    public void getTxFeeRate(GetTxFeeRateRequest req,
                             StreamObserver<GetTxFeeRateReply> responseObserver) {
        ResultHandler resultHandler = () -> {
            var reply = GetTxFeeRateReply.newBuilder()
                    .setTxFeeRateInfo(txFeeRateInfo.toProtoMessage())
                    .build();
            responseObserver.onNext(reply);
            responseObserver.onCompleted();
        };
        coreApi.getTxFeeRate(resultHandler, errorHandler);
    }
```
Bad code, just to illustrate the idea.

Reply:
This is exactly what I'm doing, will test soon...

Reply:
Commit 987d89319e8a824033879846ac43003ea219cfd0 is a solution to the blocking problem.  I don't pass an error handler to keep it simple, and not do that before I understand how it would help here.